src/controllers/settings.js

```diff
@@ -2,16 +2,198 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import AppConstants from '../app-constants.js'
+
+import {
+  getUserEmails,
+  resetUnverifiedEmailAddress,
+  addSubscriberUnverifiedEmailHash,
+  removeOneSecondaryEmail,
+  getEmailById,
+  verifyEmailHash
+} from '../db/tables/email_addresses.js'
+
+import { setAllEmailsToPrimary } from '../db/tables/subscribers.js'
+
+import { fluentError, getMessage } from '../utils/fluent.js'
+import { sendEmail, getVerificationUrl, getUnsubscribeUrl } from '../utils/email.js'
+
+import { getBreachesForEmail } from '../utils/hibp.js'
+import { generateToken } from '../utils/csrf.js'
+
 import { mainLayout } from '../views/main.js'
 import { settings } from '../views/partials/settings.js'
+import { getTemplate } from '../emails/templates/email-2022.js'
+import { verifyPartial } from '../emails/partials/email-verify.js'
+
+async function settingsPage (req, res) {
+  const emails = await getUserEmails(req.session.user.id)
+  // Add primary subscriber email to the list
+  emails.push({
+    email: req.session.user.primary_email,
+    sha1: req.session.user.primary_sha1,
+    primary: true,
+    verified: true
+  })
+
+  const breachCounts = new Map()
+
+  const allBreaches = req.app.locals.breaches
+  for (const email of emails) {
+    const breaches = await getBreachesForEmail(
+      email.sha1,
+      allBreaches,
+      true,
+      false
+    )
+    breachCounts.set(email.email, breaches?.length || 0)
+  }
 
-function settingsPage (req, res) {
   const data = {
     fxaProfile: req.user.fxa_profile_json,
-    partial: settings
+    partial: settings,
+    emails,
+    breachCounts,
+    limit: AppConstants.MAX_NUM_ADDRESSES,
+    csrfToken: generateToken(res)
   }
 
   res.send(mainLayout(data))
 }
 
-export { settingsPage }
+async function addEmail (req, res) {
+  const sessionUser = await req.user
```

Comment:
I don't understand why we need to await the `user` prop?

Reply:
Hm this came along from the bits of the user controller in V1 and I agree it's likely redundant. This is the kind of thing TS would warn about, if it knew that the type of `req` (pretty sure we could use JSDoc or other optional annotations without going full TS and get the benefits too)